pkg/registry/core/service/storage/storage.go

```diff
@@ -248,64 +247,44 @@ func (r *REST) defaultOnReadService(service *api.Service) {
 		return
 	}
 
-	if len(service.Spec.IPFamilies) > 0 {
-		return // already defaulted
-	}
+	// If IPFamilies is set, we assume it doesn't need this at all.
+	if needsClusterIP(service) && len(service.Spec.IPFamilies) == 0 {
+		singleStack := api.IPFamilyPolicySingleStack
+		requireDualStack := api.IPFamilyPolicyRequireDualStack
 
-	// set clusterIPs based on ClusterIP
-	if len(service.Spec.ClusterIPs) == 0 {
-		if len(service.Spec.ClusterIP) > 0 {
-			service.Spec.ClusterIPs = []string{service.Spec.ClusterIP}
-		}
-	}
+		if service.Spec.ClusterIP == api.ClusterIPNone {
+			// Headless.
 
-	requireDualStack := api.IPFamilyPolicyRequireDualStack
-	singleStack := api.IPFamilyPolicySingleStack
-	preferDualStack := api.IPFamilyPolicyPreferDualStack
-	// headless services
-	if len(service.Spec.ClusterIPs) == 1 && service.Spec.ClusterIPs[0] == api.ClusterIPNone {
-		service.Spec.IPFamilies = []api.IPFamily{r.primaryIPFamily}
-
-		// headless+selectorless
-		// headless+selectorless takes both families. Why?
-		// at this stage we don't know what kind of endpoints (specifically their IPFamilies) the
-		// user has assigned to this selectorless service. We assume it has dualstack and we default
-		// it to PreferDualStack on any cluster (single or dualstack configured).
-		if len(service.Spec.Selector) == 0 {
-			service.Spec.IPFamilyPolicy = &preferDualStack
-			if r.primaryIPFamily == api.IPv4Protocol {
-				service.Spec.IPFamilies = append(service.Spec.IPFamilies, api.IPv6Protocol)
-			} else {
-				service.Spec.IPFamilies = append(service.Spec.IPFamilies, api.IPv4Protocol)
-			}
-		} else {
-			// headless w/ selector
-			// this service type follows cluster configuration. this service (selector based) uses a
-			// selector and will have to follow how the cluster is configured. If the cluster is
-			// configured to dual stack then the service defaults to PreferDualStack. Otherwise we
-			// default it to SingleStack.
-			if r.secondaryIPFamily != "" {
-				service.Spec.IPFamilies = append(service.Spec.IPFamilies, r.secondaryIPFamily)
-				service.Spec.IPFamilyPolicy = &preferDualStack
+			if len(service.Spec.Selector) == 0 {
+				// Headless + selectorless is a special-case.
+				//
+				// At this stage we don't know what kind of endpoints (specifically
+				// their IPFamilies) the user has assigned to this selectorless
+				// service. We assume it has dual-stack and we default it to
+				// RequireDualStack on any cluster (single- or dual-stack
+				// configured).
+				service.Spec.IPFamilyPolicy = &requireDualStack
+				service.Spec.IPFamilies = []api.IPFamily{r.primaryIPFamily, otherFamily(r.primaryIPFamily)}
 			} else {
+				// Headless + selector - default to single.
```

Comment:
why this stops following cluster configuration? it will be incoherent with the generated endpoints.

And now that I see this, we are not using the IPFamilyPolicy field in the endpointslice controller :/

https://github.com/kubernetes/kubernetes/blob/623f9f36e00675bdf59207290f25f29cc1d24f30/pkg/controller/endpointslice/utils.go#L359-L363

Reply:
If this is a service that doesn't have ipFamilies set, it MUST have been created while the cluster was in single-stack mode.  We decided that ipFamilyPolicy was authoritative, and the user never set that.

If we make it follow the cluster, then such a service will auto-upgrade and we decided we never wanted that, right?

This is the main bug in this PR, I think.